### PR TITLE
Syncing improvements

### LINKED ISSUE #1248: Panick after running parity for a while
Got some panick error after I leave parity running for some time,

this is from running parity on testnet with --pruning fast and json-rpc and dapps enabled,

here is the backtrace

```
thread 'IO Worker #3' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Mio(Error { repr: Custom(Custom { kind: ConnectionAborted, error: StringError("Network IO notification error") }) }))', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x5620222fc9ee - core::result::unwrap_failed::h82700493d553bd18
                        at ../src/libcore/macros.rs:29
   9:     0x5620223348de - _<util..network..host..Host<Message>>::session_readable::h081dba4227a1d66a
                        at ../src/libcore/result.rs:687
                        at util/src/network/handshake.rs:131
                        at util/src/network/session.rs:210
                        at util/src/network/host.rs:638
  10:     0x56202231ebb6 - std::sys_common::unwind::try::try_fn::h9270a7ca3bcf9950
                        at util/src/io/worker.rs:103
                        at util/src/io/worker.rs:95
                        at util/src/panics.rs:84
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  11:     0x5620227523cb - __rust_try
  12:     0x56202275235d - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  13:     0x56202231ff39 - _<F as std..boxed..FnBox<A>>::call_box::hbfd3a1c90c69c5ec
                        at ../src/libstd/sys/common/unwind/mod.rs:123
                        at ../src/liballoc/boxed.rs:541
  14:     0x562022755fb4 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  15:     0x7fabfa3946f9 - start_thread
  16:     0x7fabfaf3fb5c - clone
  17:                0x0 - <unknown>
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x562022326443 - core::result::unwrap_failed::h9a5e087331deca57
                        at ../src/libcore/macros.rs:29
   9:     0x56202233808a - _<util..network..host..Host<Message> as util..IoHandler<util..NetworkIoMessage<Message>>>::update_stream::hbbadea1608576855
                        at ../src/libcore/result.rs:687
                        at util/src/network/host.rs:934
  10:     0x56202231b159 - _<util..IoManager<Message>>::start::h4c53a21c4ed59fbf
                        at util/src/io/service.rs:267
                        at /home/wighawag/.multirust/toolchains/stable/cargo/registry/src/github.com-88ac128001ac3a9a/mio-0.5.1/src/event_loop.rs:350
                        at /home/wighawag/.multirust/toolchains/stable/cargo/registry/src/github.com-88ac128001ac3a9a/mio-0.5.1/src/event_loop.rs:299
                        at /home/wighawag/.multirust/toolchains/stable/cargo/registry/src/github.com-88ac128001ac3a9a/mio-0.5.1/src/event_loop.rs:247
                        at util/src/io/service.rs:183
  11:     0x5620223188eb - std::sys_common::unwind::try::try_fn::h08d30744fb14e1f4
                        at util/src/panics.rs:84
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  12:     0x5620227523cb - __rust_try
  13:     0x56202275235d - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  14:     0x562022318e85 - _<F as std..boxed..FnBox<A>>::call_box::hd1e0445c664fc9b9
                        at ../src/libstd/sys/common/unwind/mod.rs:123
                        at ../src/liballoc/boxed.rs:541
  15:     0x562022755fb4 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  16:     0x7fabfa3946f9 - start_thread
  17:     0x7fabfaf3fb5c - clone
  18:                0x0 - <unknown>
thread 'IO Worker #1' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x562022326443 - core::result::unwrap_failed::h9a5e087331deca57
                        at ../src/libcore/macros.rs:29
   9:     0x562022327ee0 - _<util..network..host..Host<Message>>::have_session::hd90c1cb9b329095f
                        at ../src/libcore/result.rs:687
                        at ../src/libcore/iter.rs:1701
                        at util/src/network/host.rs:477
  10:     0x56202232557a - _<util..network..host..Host<Message>>::maintain_network::h7522a1d6f0ee7289
                        at ../src/libcore/iter.rs:3376
                        at ../src/libcore/iter.rs:3910
                        at util/src/network/host.rs:526
                        at util/src/network/host.rs:473
  11:     0x56202231ec36 - std::sys_common::unwind::try::try_fn::h9270a7ca3bcf9950
                        at util/src/io/worker.rs:112
                        at util/src/io/worker.rs:95
                        at util/src/panics.rs:84
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  12:     0x5620227523cb - __rust_try
  13:     0x56202275235d - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  14:     0x56202231ff39 - _<F as std..boxed..FnBox<A>>::call_box::hbfd3a1c90c69c5ec
                        at ../src/libstd/sys/common/unwind/mod.rs:123
                        at ../src/liballoc/boxed.rs:541
  15:     0x562022755fb4 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  16:     0x7fabfa3946f9 - start_thread
  17:     0x7fabfaf3fb5c - clone
  18:                0x0 - <unknown>
thread 'IO Worker #0' panicked at 'called `Result::unwrap()` on an `Err` value: "PoisonError { inner: .. }"', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x562022326443 - core::result::unwrap_failed::h9a5e087331deca57
                        at ../src/libcore/macros.rs:29
   9:     0x562022327ee0 - _<util..network..host..Host<Message>>::have_session::hd90c1cb9b329095f
                        at ../src/libcore/result.rs:687
                        at ../src/libcore/iter.rs:1701
                        at util/src/network/host.rs:477
  10:     0x56202232557a - _<util..network..host..Host<Message>>::maintain_network::h7522a1d6f0ee7289
                        at ../src/libcore/iter.rs:3376
                        at ../src/libcore/iter.rs:3910
                        at util/src/network/host.rs:526
                        at util/src/network/host.rs:473
  11:     0x56202231ec36 - std::sys_common::unwind::try::try_fn::h9270a7ca3bcf9950
                        at util/src/io/worker.rs:112
                        at util/src/io/worker.rs:95
                        at util/src/panics.rs:84
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  12:     0x5620227523cb - __rust_try
  13:     0x56202275235d - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  14:     0x56202231ff39 - _<F as std..boxed..FnBox<A>>::call_box::hbfd3a1c90c69c5ec
                        at ../src/libstd/sys/common/unwind/mod.rs:123
                        at ../src/liballoc/boxed.rs:541
  15:     0x562022755fb4 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  16:     0x7fabfa3946f9 - start_thread
  17:     0x7fabfaf3fb5c - clone
  18:                0x0 - <unknown>
thread 'IO Worker #2' panicked at 'Error deregistering stream: Io(Mio(Error { repr: Custom(Custom { kind: ConnectionAborted, error: StringError("Network IO notification error") }) }))', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x5620222fc9ee - core::result::unwrap_failed::h82700493d553bd18
                        at ../src/libcore/macros.rs:29
   9:     0x562022327419 - _<util..network..host..Host<Message>>::kill_connection::hf378240737f1b781
                        at ../src/libcore/result.rs:708
                        at util/src/network/host.rs:745
  10:     0x562022329d84 - _<util..network..host..Host<Message> as util..IoHandler<util..NetworkIoMessage<Message>>>::timeout::hbbf100bff0d462ec
                        at util/src/network/host.rs:707
                        at util/src/network/host.rs:819
  11:     0x56202231ec36 - std::sys_common::unwind::try::try_fn::h9270a7ca3bcf9950
                        at util/src/io/worker.rs:112
                        at util/src/io/worker.rs:95
                        at util/src/panics.rs:84
                        at ../src/libstd/sys/common/unwind/mod.rs:127
  12:     0x5620227523cb - __rust_try
  13:     0x56202275235d - std::sys_common::unwind::inner_try::h9eebd8dc83f388a6
  14:     0x56202231ff39 - _<F as std..boxed..FnBox<A>>::call_box::hbfd3a1c90c69c5ec
                        at ../src/libstd/sys/common/unwind/mod.rs:123
                        at ../src/liballoc/boxed.rs:541
  15:     0x562022755fb4 - std::sys::thread::Thread::new::thread_start::h471ad90789353b5b
  16:     0x7fabfa3946f9 - start_thread
  17:     0x7fabfaf3fb5c - clone
  18:                0x0 - <unknown>
2016-06-08 19:21:30 UTCINFO:parity: Finishing work, please wait...
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: NotifyError::Closed(..)', ../src/libcore/result.rs:746
stack backtrace:
   1:     0x5620227531f0 - std::sys::backtrace::tracing::imp::write::h4c73fcd3363076f5
   2:     0x56202275794b - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::h0422dbb3077e6747
   3:     0x5620227575d3 - std::panicking::default_hook::haac48fa641db8fa2
   4:     0x5620227404af - std::sys_common::unwind::begin_unwind_inner::h39d40f52add53ef7
   5:     0x5620227414e8 - std::sys_common::unwind::begin_unwind_fmt::h64c0ff793199cc1b
   6:     0x562022752441 - rust_begin_unwind
   7:     0x56202278e3df - core::panicking::panic_fmt::h73bf9d7e8e891a73
   8:     0x562021f77751 - core::result::unwrap_failed::h0e8a7f0e22819924
                        at ../src/libcore/macros.rs:29
   9:     0x562021f77449 - ethcore..service..ClientService::drop.21663::ha2c96f03a2e5a06b
  10:     0x562021f71b36 - parity::execute_client::h20d1f30272432502
                        at parity/main.rs:266
  11:     0x562021f5f799 - parity::execute::h4228f2b3aef67c46
                        at parity/main.rs:145
  12:     0x562021f5cb9f - parity::main::hc0b8b9a4c77213f8
                        at parity/main.rs:103
  13:     0x562022757214 - std::sys_common::unwind::try::try_fn::h09ba69fd13531e58
  14:     0x5620227523cb - __rust_try
  15:     0x562022756c5b - std::rt::lang_start::h5b0863080165c75e
  16:     0x7fabfae5982f - __libc_start_main
  17:     0x562021f5c2b8 - _start
  18:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by arkpar**
Fixes #1248 and #1228

### CODE REVIEW

File: util/src/network/session.rs [hunk 1]
```diff
@@ -329,16 +331,19 @@ impl Session {
 			PACKET_DISCONNECT => {
 				let rlp = UntrustedRlp::new(&packet.data[1..]);
 				let reason: u8 = try!(rlp.val_at(0));
+				if self.had_hello {
```
Comment (gavofyork) [inline util/src/network/session.rs]: what's this conditional about?

File: sync/src/chain.rs [hunk 2]
```diff
@@ -1116,10 +1145,10 @@ impl ChainSync {
 						let mut rlp_stream = RlpStream::new_list(blocks.len());
 						for block_hash in  blocks {
 							let mut hash_rlp = RlpStream::new_list(2);
-							let difficulty = chain.block_total_difficulty(BlockID::Hash(block_hash.clone())).expect("Malformed block without a difficulty on the chain!");
+							let number = HeaderView::new(&chain.block_header(BlockID::Hash(block_hash.clone())).expect("Malformed block without a header on the chain!")).number();
```
Comment (gavofyork) [inline sync/src/chain.rs]: `expect` text should include some explicit proof that `block_hash` is definitely one of our imported blocks.

File: sync/src/chain.rs [hunk 3]
```diff
@@ -831,6 +843,9 @@ impl ChainSync {
 		rlp.append(&skip);
 		rlp.append(&if reverse {1u32} else {0u32});
 		self.send_request(sync, peer_id, asking, GET_BLOCK_HEADERS_PACKET, rlp.out());
+		if let Some(ref mut peer) = self.peers.get_mut(&peer_id) {
```
Comment (gavofyork) [inline sync/src/chain.rs]: reason for contional?

File: sync/src/chain.rs [hunk 4]
```diff
@@ -441,6 +449,10 @@ impl ChainSync {
 			}
 		}
 
+		if !valid_response {
```
Comment (gavofyork) [inline sync/src/chain.rs]: reason for contional?

File: sync/src/chain.rs [hunk 5]
```diff
@@ -408,9 +410,15 @@ impl ChainSync {
 
 		let mut headers = Vec::new();
 		let mut hashes = Vec::new();
+		let mut valid_response = false;
 		for i in 0..item_count {
 			let info: BlockHeader = try!(r.val_at(i));
 			let number = BlockNumber::from(info.number);
+			if !valid_response {
```
Comment (gavofyork) [inline sync/src/chain.rs]: reason for contional?

File: util/src/network/session.rs (hunk 1 above)
Comment (arkpar) [inline util/src/network/session.rs]: This is to filter out disconnects that happen before the session is established. Disconnect that happens when the session is active is much more interesting to log as it means that the peer decided to disconnect us based on our activity

Comment (arkpar) [inline util/src/network/session.rs]: All disconnects are additionally logged on the `trace` level in host.rs

File: sync/src/chain.rs [hunk 6]
```diff
@@ -921,13 +936,24 @@ impl ChainSync {
...
 			// id is a hash
 			let hash: H256 = try!(r.val_at(0));
-			trace!(target: "sync", "-> GetBlockHeaders (hash: {}, max: {}, skip: {}, reverse:{})", hash, max_headers, skip, reverse);
+			trace!(target: "sync", "{} -> GetBlockHeaders (hash: {}, max: {}, skip: {}, reverse:{})", peer_id, hash, max_headers, skip, reverse);
 			match io.chain().block_header(BlockID::Hash(hash)) {
-				Some(hdr) => From::from(HeaderView::new(&hdr).number()),
+				Some(hdr) => {
+					let number = From::from(HeaderView::new(&hdr).number());
+					assert_eq!(HeaderView::new(&hdr).sha3(), hash);
+					if max_headers == 1 || io.chain().block_hash(BlockID::Number(number)) != Some(hash) {
```
Comment (gavofyork) [inline sync/src/chain.rs]: can handle single-step reverse hashchains of non-canon hashes, too...

**Pull request merged**